### PR TITLE
Breaking change: u-root works exactly when go build or go list also work.

### LINKED ISSUE #985: binary build mode fails when including commands with vendor dependencies
Generating an image with "binary" build mode fails:
```
git log  --pretty=oneline --abbrev-commit   -1
8ebf81a3 (HEAD -> master, origin/master, origin/HEAD) initramfs: now that we follow symlinks, don't lstat.

./u-root -defaultsh rush -build binary -o ./myinitrd github.com/u-root/u-root/cmds/{init,ls} github.com/u-root/u-root/xcmds/rush
2018/10/18 15:44:58 Disabling CGO for u-root...
2018/10/18 15:44:58 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/usr/lib64/go/1.11 GOPATH=/home/ddiss/go:/usr/share/go/1.11/contrib CGO_ENABLED=0
2018/10/18 15:44:58 Filename is ./myinitrd
2018/10/18 15:45:01 error building builder.Opts{Env:golang.Environ{Context:build.Context{GOARCH:"amd64", GOOS:"linux", GOROOT:"/usr/lib64/go/1.11", GOPATH:"/home/ddiss/go:/usr/share/go/1.11/contrib", CgoEnabled:false, UseAllFiles:false, Compiler:"gc", BuildTags:[]string(nil), ReleaseTags:[]string{"go1.1", "go1.2", "go1.3", "go1.4", "go1.5", "go1.6", "go1.7", "go1.8", "go1.9", "go1.10", "go1.11"}, InstallSuffix:"", JoinPath:(func(...string) string)(nil), SplitPathList:(func(string) []string)(nil), IsAbsPath:(func(string) bool)(nil), IsDir:(func(string) bool)(nil), HasSubdir:(func(string, string) (string, bool))(nil), ReadDir:(func(string) ([]os.FileInfo, error))(nil), OpenFile:(func(string) (io.ReadCloser, error))(nil)}}, Packages:[]string{"github.com/u-root/u-root/cmds/init", "github.com/u-root/u-root/cmds/ls", "github.com/u-root/u-root/xcmds/rush"}, TempDir:"/tmp/u-root542580345/builder429597060", BinaryDir:"bin"}: error building go package in "/home/ddiss/go/src/github.com/u-root/u-root/cmds/ls": ls.go:27:2: cannot find package "github.com/spf13/pflag" in any of:
        /usr/lib64/go/1.11/src/github.com/spf13/pflag (from $GOROOT)
        /home/ddiss/go/src/github.com/spf13/pflag (from $GOPATH)
        /usr/share/go/1.11/contrib/src/github.com/spf13/pflag
, exit status 1
```
The missing _github.com/spf13/pflag_ dependency is present in the top level vendor dir:
```
ls -d vendor/github.com/spf13/pflag
vendor/github.com/spf13/pflag
```

Using "bb" build mode works fine:
```
./u-root -defaultsh rush -build bb -o ./myinitrd github.com/u-root/u-root/cmds/{init,ls} github.com/u-root/u-root/xcmds/rush2018/10/18 15:41:47 Disabling CGO for u-root...
2018/10/18 15:41:47 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/usr/lib64/go/1.11 GOPATH=/home/ddiss/go:/usr/share/go/1.11/contrib CGO_ENABLED=0
2018/10/18 15:41:47 Filename is ./myinitrd
2018/10/18 15:41:52 Successfully wrote initramfs.
```

### LINKED ISSUE #1024: Builder does not use vendoring
Hi,

When building u-bmc without an u-root checked out I get this:

```
GOARM=5 GOARCH=arm ./u-bmc -o "initramfs.cpio.tmp" -p "quanta-f06-leopard-ddr3"
2018/11/13 16:17:30 Build environment: GOARCH=arm GOOS=linux GOROOT=/usr/local/go GOPATH=/home/bluecmd/go CGO_ENABLED=0
2018/11/13 16:17:30 Filename is initramfs.cpio.tmp
2018/11/13 16:17:30 "github.com/u-root/u-root/cmds/ansi" is neither package or path/glob: cannot find package "github.com/u-root/u-root/cmds/ansi" in any of:
    /usr/local/go/src/github.com/u-root/u-root/cmds/ansi (from $GOROOT)
    /home/bluecmd/go/src/github.com/u-root/u-root/cmds/ansi (from $GOPATH)
```

Note that it does not try to look at u-bmc's vendor directory.

As a workaround I tried `ln -sf ~/go/src/github.com/u-root/u-bmc/vendor/github.com/u-root/u-root ~/go/src/github.com/u-root/u-root`, and that seems to work for now.

### LINKED ISSUE #1851: Inconsistent versions when getting master with modules enabled
If I run the steps below, I get an error about inconsistent versions:
```
mkdir test
cd test
go mod init test
go get -u github.com/u-root/u-root@master
```

Gives me the error:
```
go: github.com/u-root/u-root master => v1.0.1-0.20201001202801-7709c8272b97
go: golang.org/x/tools upgrade => v0.0.0-20201010145503-6e5c6d77ddcc
go: golang.org/x/xerrors upgrade => v0.0.0-20200804184101-5ec99f83aff1
go: golang.org/x/sys upgrade => v0.0.0-20201009025420-dfb3f7c4e634
go: golang.org/x/mod upgrade => v0.3.0
go get: inconsistent versions:
	github.com/u-root/u-root@v1.0.1-0.20201001202801-7709c8272b97 from github.com/u-root/u-root@master requires github.com/u-root/u-root@v6.0.1-0.20200118052101-6bcd1cda5996+incompatible (not github.com/u-root/u-root@v1.0.1-0.20201001202801-7709c8272b97 from github.com/u-root/u-root@master)
```

### LINKED ISSUE #1863: Documentation: examples using curly braces globs don't work in practice.
The README suggests using {a, b, c} to install commands from the exp/ category. This however doesn't work because go's filepath package does not support curly braces globbing, as shown here: https://golang.org/pkg/path/filepath/#Match

### LINKED ISSUE #2149: u-root busybox mode is not working with some go modules
Hello, 
Impossible to use some external module in bb mode, 
example : 
https://github.com/hanwen/go-fuse/

=> 
 could not import github.com/hanwen/go-fuse/v2/fs (cannot find package "github.com/hanwen/go-fuse/v2/fs" in any of:
        /usr/lib/go/src/github.com/hanwen/go-fuse/v2/fs (from $GOROOT)
        /tmp/workdir/GOLANG/src/github.com/hanwen/go-fuse/v2/fs (from $GOPATH))))

perhaps because 
https://github.com/hanwen/go-fuse/blob/master/go.mod    
that name /v2  and not the root of the github .

Regards, 
Nicolas

### LINKED ISSUE #2198: Cannot makebb for dependency of gdamore/tcell
This script duplicates the error message "could not import github.com/gdamore/tcell (type-checking package "github.com/gdamore/tcell" failed"
```
#!/bin/bash

mkdir bug
cd bug
mkdir -p src/simple
mkdir -p src/hello
cat <<EOF > src/simple/main.go
package main

import (
	"fmt"
	"github.com/gdamore/tcell"
)

func main() {
	fmt.Println(tcell.NewScreen)
}
EOF

cat <<EOF > src/hello/main.go
package main

import "fmt"

func main() {
	fmt.Println("hello")
}
EOF
GOPATH="$HOME/bugbb:$HOME/go" GO111MODULE=off makebb hello 
GOPATH="$HOME/bugbb:$HOME/go" GO111MODULE=off makebb hello simple
```

**Additional context**
This used to work it seems, so maybe this is related to go modules being broken.

### LINKED ISSUE #2200: Cannot build u-root in Ubuntu 18.04 with go1.17.3
**Describe the bug**
Running "go get github.com/u-root/u-root" using go1.17.3 causes u-root to fail

**To Reproduce**
```
# I have limited space under $HOME so I have to set GOPATH and GOCACHE
$ mkdir -p /local/mnt2/workspace2/go
$ mkdir -p /local/mnt2/workspace2/gocache
$ export GOPATH=/local/mnt2/workspace2/go
$ export GOCACHE=/local/mnt2/workspace2/gocache
# Install go1.17.3 at /usr/local/go
$ export PATH=/usr/local/go/bin
$ git clone https://github.com/u-root/u-root
$ cd u-root
$ go get github.com/u-root/u-root
$ u-root
2021/12/01 14:23:50 Disabling CGO for u-root...
2021/12/01 14:23:50 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/usr/local/go GOPATH=/local/mnt2/workspace2/go CGO_ENABLED=0
2021/12/01 14:23:50 "github.com/u-root/u-root/cmds/core/*" is neither package or path/glob: malformed import path "github.com/u-root/u-root/cmds/core/*": invalid char '*'
```

I can "fix" this by reverting to tag v0.7.0 and running "go get github.com/u-root/u-root" and u-root using go1.14.15, then switching to master.

```
$ sudo rm -rf /local/mnt2/workspace2/gocache/* /local/mnt2/workspace2/go/*
$ git checkout v0.7.0
$ GOROOT=/opt/go1.14/go PATH=/opt/go1.14/go/bin:$PATH go get github.com/u-root/u-root
$ GOROOT=/opt/go1.14/go PATH=/opt/go1.14/go/bin:$PATH u-root
2021/12/01 14:39:05 Disabling CGO for u-root...
2021/12/01 14:39:05 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/opt/go1.14/go GOPATH=/local/mnt2/workspace2/go CGO_ENABLED=0
2021/12/01 14:39:05 WARNING: You are not using one of the recommended Go versions (have = go1.14.15, recommended = [go1.17]).
                        Some packages may not compile.
                        Go to https://golang.org/doc/install to find out how to install a newer version of Go,
                        or use https://godoc.org/golang.org/dl/go1.17 to install an additional version of Go.
2021/12/01 14:39:05 Skipping package "/local/mnt2/workspace2/go/src/github.com/u-root/u-root/cmds/core/bind": no buildable Go source files in /local/mnt2/workspace2/go/src/github.com/u-root/u-root/cmds/core/bind
2021/12/01 14:39:11 Successfully built "/tmp/initramfs.linux_amd64.cpio" (size 15895832).
$ git checkout master
$ go get github.com/u-root/u-root
$ u-root
2021/12/01 14:40:13 Disabling CGO for u-root...
2021/12/01 14:40:13 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/usr/local/go GOPATH=/local/mnt2/workspace2/go CGO_ENABLED=0
2021/12/01 14:40:13 Skipping package "/local/mnt2/workspace2/go/src/github.com/u-root/u-root/cmds/core/bind": no buildable Go source files in /local/mnt2/workspace2/go/src/github.com/u-root/u-root/cmds/core/bind
2021/12/01 14:40:16 Successfully built "/tmp/initramfs.linux_amd64.cpio" (size 12524824).
```

**Expected behavior**
I would expect that it works with go1.17.3

### LINKED ISSUE #2450: u-root builds should fail and do not, creating initramfs with missing commands
**Describe the bug**
I found that even when cpud had compile errors, the u-root tool would create an initramfs with no error. This must be fixed.

**To Reproduce**
```
rminnich@a300:~/go/src/github.com/u-root/u-root$ echo bad >>cmds/core/date/date.go
rminnich@a300:~/go/src/github.com/u-root/u-root$ ./u-root
17:39:21 Disabling CGO for u-root...
17:39:21 Build environment: GOARCH=amd64 GOOS=linux GOROOT=/usr/lib/go-1.18 GOPATH=/home/rminnich/go CGO_ENABLED=0
17:39:21 WARNING: You are not using one of the recommended Go versions (have = go1.18.1, recommended = [go1.17]).
			Some packages may not compile.
			Go to https://golang.org/doc/install to find out how to install a newer version of Go,
			or use https://godoc.org/golang.org/dl/go1.17 to install an additional version of Go.
17:39:21 NOTE: building with the new gobusybox; to get the old behavior check out commit 8b790de
17:39:21 Skipping package "cmds/core/bind": no buildable Go source files in /home/rminnich/go/src/github.com/u-root/u-root/cmds/core/bind
17:39:21 Skipping package "cmds/core/installcommand": no buildable Go source files in /home/rminnich/go/src/github.com/u-root/u-root/cmds/core/installcommand
17:39:21 Skipping package "cmds/core/timeout": no buildable Go source files in /home/rminnich/go/src/github.com/u-root/u-root/cmds/core/timeout
/home/rminnich/go/src/github.com/u-root/u-root/cmds/core/date/date.go:263:1: expected declaration, found bad
17:39:30 Successfully built "/tmp/initramfs.linux_amd64.cpio" (size 12553752).
rminnich@a300:~/go/src/github.com/u-root/u-root$ ls -l /tmp/initramfs.linux_amd64.cpio
-rw-r--r-- 1 rminnich rminnich 12553752 Jul 25 17:39 /tmp/initramfs.linux_amd64.cpio
```

**Expected behavior**
Well, for sure, a failed build should not result in "Successfully built." This has the potential to brick every system using u-root for firmware. I am at a loss to understand how this ever got in.

### LINKED ISSUE #2603: build directory is removed even though it says it is being preserved
**Describe the bug**
on error, u-root says it is preserving the directory when it is not

**To Reproduce**
./u-root -go-extra-args -compiler=/usr/local/src/tinygo/bin/tinygo cmds/core/date
21:18:59 Disabling CGO for u-root...
21:18:59 Build environment: GOARCH=amd64 GOOS=linux GOPATH=/usr/local/google/home/rminnich/go CGO_ENABLED=0 GO111MODULE= GOROOT=/usr/local/src/go PATH=/us (lost due to wrapping)
21:18:59 NOTE: building with the new gobusybox; to get the old behavior check out commit 8b790de21:19:00 Build error: error building: preserving bb generated source directory at /tmp/u-root1853703085/builder3556591098 due to error. To debug build, `c
go build with modules failed: error building go package in "/tmp/u-root1853703085/builder3556591098/src/bb.u-root.com/bb": invalid value "/usr/local/src/t
usage: go build [-o output] [build flags] [packages]
Run 'go help build' for details.
, exit status 2
rminnich@glenda:~/go/src/github.com/u-root/u-root$ pushd /tmp/u-root1853703085/builder3556591098
-bash: pushd: /tmp/u-root1853703085/builder3556591098: No such file or directory

rminnich@glenda:~/go/src/github.com/u-root/u-root$ ls -l /tmp/u-root1853703085
ls: cannot access '/tmp/u-root1853703085': No such file or directory
rminnich@glenda:~/go/src/github.com/u-root/u-root$ 

**Expected behavior**
the directory is there

**Additional context**
trying to use tinygo as the compiler.

### LINKED ISSUE #2814: u-root lies about saving tmp dir
**Describe the bug**
u-root says it is preserving the generated source directory, but it is not

./u-root -gen-dir  ~/tmp/whatever -tags tinygo cmds/exp/rush

13:21:46 Build error: error building: preserving bb generated source directory at /var/folders/vp/n2xq_79s3s549qb4tt4380v00000gp/T/u-root2050172721/builder265485357 due to error:
finding packages failed: could not load packages from file system: failed to add package github.com/u-root/u-root/cmds/exp/rush for errors: 1 error occurred:
	* /Users/rminnich/go/src/github.com/u-root/u-root/cmds/exp/rush/rush.go:95:9: undefined: runone

ronsexcllentmbp:u-root rminnich$ ls /var/folders/vp/n2xq_79s3s549qb4tt4380v00000gp/T/u-root*

ls: /var/folders/vp/n2xq_79s3s549qb4tt4380v00000gp/T/u-root*: No such file or directory

**To Reproduce**
./u-root -gen-dir  ~/tmp/whatever -tags tinygo cmds/exp/rush

**Expected behavior**
I expect it to do what it says it is doing.
13:21:46 Build error: error building: preserving bb generated source directory at /var/folders/vp/n2xq_79s3s549qb4tt4380v00000gp/T/u-root2050172721/builder265485357 due to error:

**Additional context**
It worked as of last february, and is broken again.

### PR DESCRIPTION
**Pull request by hugelgupf**
> [!IMPORTANT]
> TL;DR: You won't be able to run `u-root` from any arbitrary directory after this change (when using `GO111MODULE=on|auto`). `u-root` must be run somewhere with a Go module or Go workspace.
>
> Go workspaces replace this use case, and there is a tool to create Go workspaces on the fly called [goanywhere](https://github.com/u-root/gobusybox?tab=readme-ov-file#goanywhere). Another usable method is described by the [mkuimage README](https://github.com/u-root/mkuimage?tab=readme-ov-file#multi-module-workspace-builds).

This change pulls in https://github.com/u-root/gobusybox/pull/110, in which we stop supporting module builds without a go.mod. In essence, from this point forward, the `u-root` tool will support builds exactly when `go build` and `go list` also work. https://github.com/u-root/gobusybox/pull/110 reduces a lot of maintenance burden, trivially starts supporting Go workspaces and Go module replace directives, and increases gobusybox code's readability.

Multi-module builds can be done easily with the 2 standard Go methods, as described in the [mkuimage README](https://github.com/u-root/mkuimage?tab=readme-ov-file#multi-module-workspace-builds). Go workspaces, vendored Go workspaces, Go modules, and vendored Go modules are all supported. Completely offline builds can easily done with vendored Go modules or workspaces.

For ease of use, the [goanywhere](https://github.com/u-root/gobusybox?tab=readme-ov-file#goanywhere) tool was created. `goanywhere` creates a temporary directory with a Go workspace of the given commands, and then execs a binary passing along the Go command paths. Use like `goanywhere ./u-root/cmds/core/{init,gosh} ./cpu/cmds/cpud -- u-root [other u-root args]`. This approaches the easy usability of `u-root` up to this point.

`goanywhere` does not work with templates. `goanywhere` only accepts file system paths at this time.

This also fixes the issue where `binary` mode builds didn't always work in the same cases where `gbb` builds worked. Both are now always supported in the exact same circumstances.

This change also pulls in https://github.com/u-root/mkuimage/pull/28, in which we support generic template YAML files named ".mkuimage.yaml" in the current working directory or any of its parents. Templates support the existing command expansions, but also support configs (invoked with `u-root -config=$config`) in which one can specify build configuration and mixed-builder (bb/binary) builds. They will be documented in a README at a later time.

Also check out the new [README](https://github.com/hugelgupf/u-root/blob/mkuimage-cfg/README.md).

Closes #2814 
Closes #2603 
Closes #2450 
Closes #2200 
Closes #2198 
Closes #2149
Closes #1863
Closes #1851 
Closes #1024
Closes #985

### CODE REVIEW

**Review by rminnich — APPROVED** [review]
nice

**Review by rminnich — DISMISSED** [review]

**Comment by hugelgupf** [comment]
I'll set up some more tools in the other repo before I really merge this -- I'd like e.g. an easy way to test compilation of every template.

**Comment by hugelgupf** [comment]
PTAL

**Comment by MDr164** [comment]
Really nice! It's fairly large so looking over everything might take time but so far this looks very promising. The gist is to strip out the image building and rely on mkuimage and gobusybox alone, right? Seems reasonable to me.

**Comment by hugelgupf** [comment]
> Really nice! It's fairly large so looking over everything might take time but so far this looks very promising. The gist is to strip out the image building and rely on mkuimage and gobusybox alone, right? Seems reasonable to me.

That is a side-effect here. The behavior changes (which are pulled in through that) are important to take a look at.

**Review by rminnich — APPROVED** [review]
this makes a lot of sense. the nature of u-root is utterly changed from 12y ago, as is the nature of Go, and this seems like a long overdue reworking of it all.
